library/core/src/iter/traits/iterator.rs

```diff
@@ -2998,6 +2998,11 @@ pub trait Iterator {
     /// This is useful when you have an iterator over `&T`, but you need an
     /// iterator over `T`.
     ///
+    /// Note that unlike copying, cloning is regarded as a side effect, because
+    /// `Clone` implementations may contain observable behavior, so for example
+    /// on `iter.cloned().last()`, all elements will be cloned even if only the
+    /// last one is actually returned.
```

Comment:
Is it really worth making this guarantee? I could imagine a future compiler optimization (or perhaps specialization) that would eliminate `clone` calls for non–side effecting `Clone` implementors, such as `Vec`.

Reply:
We will soon have a clippy lint for this, so no, I don't think we should sacrifice backwards compatibility for a possible perf gain.

Reply:
Maybe the clippy lint is the issue then?

Reply:
Also, there is no backwards compatibility issue since the API contract currently does not promise this. It only becomes an issue once the guarantee is added.

Reply:
There is no "promise", but it has been observable behavior since 1.0.0. So if we were to change that, we better have a good idea if any code depends on it.

Reply:
Also why would the clippy lint be an issue?

Reply:
I assume the clippy lint is implying something about the side-effects since it was brought up here and since the lint is motivating this change? The lint itself wasn't linked, so I am just guessing.

> There is no "promise", but it has been observable behavior since 1.0.0. 

observable behavior != stability guarantee

This PR turns what someone might *hypothetically* rely on into a guarantee, that's a significant change that paints us into a corner.

Reply:
I'm not wedded to this PR – however I'd like to have the (non-)guarantees documented. I'd swap this out with "you should not rely on the items' `Clone` impl being called for all items observed by the iterator" in a heartbeat if that's what we decide.

Leaving the observable behavior undocumented makes it harder to say "you shouldn't have relied on it" later.